File: public/dsa/linked_lists/delete_xth_node_from_end_of_linked_list/approach/delete_xth_node_from_end_of_linked_list.cpp
#include <bits/stdc++.h>
using namespace std;

// defs-start
struct ListNode {
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};
// defs-end

ListNode* removeXthNodeFromEnd(ListNode* head, int x) {
    ListNode* dummy = new ListNode(0);
    dummy->next = head;
    ListNode* fast = dummy;
    ListNode* slow = dummy;

    for (int i = 0; i < x + 1; i++) {
        if (!fast) return head;
        fast = fast->next;
    }

    while (fast) {
        fast = fast->next;
        slow = slow->next;
    }

    ListNode* nodeToDelete = slow->next;
    slow->next = slow->next->next;
    delete nodeToDelete;

    return dummy->next;
}

// tests-start
void printList(ListNode* head) {
    while (head) {
        cout << head->data;
        if (head->next) cout << " ";
        head = head->next;
    }
    cout << endl;
}

int main() {
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);

    int x = 2;
    head = removeXthNodeFromEnd(head, x);

    printList(head);
    return 0;
}
// tests-end

/*output
1 2 4
*/

File: public/dsa/linked_lists/delete_xth_node_from_end_of_linked_list/approach/delete_xth_node_from_end_of_linked_list.py
# defs-start
class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def removeXthNodeFromEnd(self, head: ListNode, x: int) -> ListNode:
        # Create a dummy node to handle edge cases (like deleting the head)
        dummy = ListNode(0, head)
        slow = dummy
        fast = dummy

        # Move fast x+1 steps ahead so that slow points to node before target
        for _ in range(x + 1):
            if not fast:
                return head  # x > length of list
            fast = fast.next

        # Move both pointers until fast reaches the end
        while fast:
            fast = fast.next
            slow = slow.next

        # Delete target node
        slow.next = slow.next.next

        return dummy.next


# tests-start
def printList(node):
    while node:
        print(node.data, end=" " if node.next else "\n")
        node = node.next


head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
sol = Solution()
new_head = sol.removeXthNodeFromEnd(head, 2)
printList(new_head)
# tests-end

'''output
1 2 4
'''
